packages/common/src/state/simple.ts

```diff
@@ -0,0 +1,185 @@
+    }
+    this.accountStack.push(newTopA)
+    this.codeStack.push(new Map(this.topC()))
+    this.storageStack.push(new Map(this.topS()))
+  }
+
+  async getAccount(address: Address): Promise<Account | undefined> {
+    return this.topA().get(address.toString())
+  }
+
+  async putAccount(address: Address, account?: Account | undefined): Promise<void> {
+    this.topA().set(address.toString(), account)
+  }
+
+  async deleteAccount(address: Address): Promise<void> {
+    this.topA().set(address.toString(), undefined)
+  }
+
+  async modifyAccountFields(address: Address, accountFields: AccountFields): Promise<void> {
+    let account = await this.getAccount(address)
+    if (!account) {
+      account = new Account()
+    }
+    account.nonce = accountFields.nonce ?? account.nonce
+    account.balance = accountFields.balance ?? account.balance
+    account.storageRoot = accountFields.storageRoot ?? account.storageRoot
+    account.codeHash = accountFields.codeHash ?? account.codeHash
+    await this.putAccount(address, account)
+  }
+
+  async getContractCode(address: Address): Promise<Uint8Array> {
+    return this.topC().get(address.toString()) ?? new Uint8Array(0)
+  }
+
+  async putContractCode(address: Address, value: Uint8Array): Promise<void> {
+    this.topC().set(address.toString(), value)
+    if ((await this.getAccount(address)) === undefined) {
+      await this.putAccount(address, new Account())
+    }
+    await this.modifyAccountFields(address, { codeHash: keccak256(value) })
+  }
+
+  async getContractCodeSize(address: Address): Promise<number> {
+    const contractCode = await this.getContractCode(address)
+    return contractCode.length
+  }
+
+  async getContractStorage(address: Address, key: Uint8Array): Promise<Uint8Array> {
+    return this.topS().get(`${address.toString()}_${bytesToHex(key)}`) ?? new Uint8Array(0)
+  }
+
+  async putContractStorage(address: Address, key: Uint8Array, value: Uint8Array): Promise<void> {
+    this.topS().set(`${address.toString()}_${bytesToHex(key)}`, value)
+  }
+
+  async checkpoint(): Promise<void> {
+    this.add()
+  }
+  async commit(): Promise<void> {
+    const topA = this.accountStack.pop()
+    const topC = this.codeStack.pop()
+    const topS = this.storageStack.pop()
+    this.accountStack.pop()
+    this.codeStack.pop()
+    this.storageStack.pop()
+    this.accountStack.push(topA!)
+    this.codeStack.push(topC!)
+    this.storageStack.push(topS!)
+  }
+
+  async revert(): Promise<void> {
+    this.accountStack.pop()
+    this.codeStack.pop()
+    this.storageStack.pop()
+  }
+
+  async flush(): Promise<void> {}
+  clearCaches(): void {}
+
+  shallowCopy(): EVMStateManagerInterface {
+    const copy = new SimpleStateManager()
+    for (let i = 0; i < this.accountStack.length; i++) {
+      copy.accountStack.push(new Map(this.accountStack[i]))
+      copy.codeStack.push(new Map(this.codeStack[i]))
+      copy.storageStack.push(new Map(this.storageStack[i]))
```

Comment:
So this shallow copy will initialize the shallow copied state manager with the current state of the stack of this state manager.

For the `accountStack`, if you edit the account on the original state manager, then it will also be edited on the copied state manager, I do not think that this is desired? (So need to create copies of the accounts of the `accountStack`).

Side-note: if you would raw-edit the bytes on the `codeStack` or the `storageStack` items these will also be changed in both state managers. However I don't think we ever edit raw bytes so this should not be a problem.

Reply:
> For the accountStack, if you edit the account on the original state manager, then it will also be edited on the copied state manager, I do not think that this is desired?

This is the definition of `shallowCopy()` and that's also the behavior from the `DefaultStateManager` and that's why we introduced this new name, to align with this behavior.

So if we want a full copy method, we should likely (?) call this `copy()` and do in addition. This would draw in interface changes though, so I would think we should not do here. My goal here was to get as close to the existing implementation as possible.

Reply:
Ah right, I already knew I was confused, you are right regarding the shallow copy, so we can keep. We should indeed not deep copy.

Reply:
Yeah, but I honestly also really was confused first round to somewhat be forced to add the worse option here (we introduced `shallowCopy()` because for a full-blow level DB state it's just not realistic to have a full copy, but in most use cases a full copy would actually be *better*).

I guess what we can do (mid term) is that we add a `copy()` method to the EVM/VM additionally to the now `shallowCopy()` implementations which then checks if the current active `StateManager` and `Blockchain` instances have `copy()` methods (we would e.g. for this SM then also add an additional `copy()`) method and if they do `copy()` is done and if not an Error is thrown.

Guess that would be a pretty nice additional functionality for use cases where people *want* a full copy (of the EVM e.g.), e.g. to have isolated test runs or whatever.